proposals/2025-12-02_controller_shared_data.md

```diff
@@ -0,0 +1,215 @@
+has separate devices for the controller and the two screens (two total USB
+devices).
+
+## Non-Goals
+
+* We do not intend to fully support the multi-device scenario yet, that needs
+  further design to associate specific devices with specific controller
+  configurations.
+* We do not intend to immediately support "global" objects accessible across
+  namespaces, like "universal shift".
+
+## "Universal Shift"
+
+"Universal Shift" refers to the idea that a shift button pressed on one
+controller can be detected by any and all other controllers. This may be a
+useful use-case but creates a lot of difficulties, so for now Universal Shift is
+out of scope for this first implementation.
+
+## How
+
+### Data Object
+
+We will create a central object inside Mixxx that contains a triple-keyed map:
+
+* Namespace (string)
+  * Entity (string)
+    * Key (string)
+      * Value ("SafeData")
+
+#### Namespace
+
+`Namespace` is a string that is unique to each controller **mapping
+definition**. All connected controllers of the same model will share the same
+namespace. e.g. Two CDJ-2000's will both have a namespace like `CDJ_2000`. No
+two hardware mappings will have the same namespace, and we can enforce that with
+a precommit check.
+
+#### Entity
+
+`Entity` is a logical value defined by the controller mapping definition. It
+can be like a Mixxx-style group ("`[Channel1]`") but it can be any arbitrary
+string. Many controllers will want to define something like `deck1` to refer to
+a device that can itself be assigned to multiple mixxx channels.
+
+The controller mapping decides how these entities behave and Mixxx does no
+enforcement of them. To reiterate: even if an "entity" "looks like" a Mixxx
+group, it is not.
+
+#### Key
+
+`Key` is a logical value defined by the controller mapping definition. It could
+refer to a button, light, knob, or abstract name.
+
+The controller mapping decides how these keys behave and Mixxx does no
+enforcement of them. Similar to "entity", keys bear no relation to equivalent
+Mixxx keys.
+
+#### Value
+
+In the engine, the `value` is stored as a QJSValue, however we want to only
+support a limited set of types in Javascript / Typescript. For the first
+implementation, we will support bool, number, and string, and Arrays of those:
+
+```typescript
+type SafePrimitive = string | number | boolean | null;
+
+type SafeData =
+  | SafePrimitive
+  | SafePrimitive[];
+```
+
+The list of allowable types can be expanded as needed, but we want to be sure
+that the shared data system does not become a "bag of bytes" message bus for
+large pieces of data like bitmaps or code, nor should it be used to circumvent
+intentional limitations or gaps in the overall javascript framework.
+
+#### Example
+
+The shift button on the left side of a Traktor S4MK3 would be stored this way:
+
+pseudocode:
+
+`sharedData["S4MK3"]["deck1"]["shift"] = true`
+
+### API
```

Comment:
Currently, the rest of the `engine` [API](https://github.com/mixxxdj/mixxx/blob/main/res/controllers/engine-api.d.ts#L118) uses `lowerCamelCase`. Could you please update the current `PascalCase` function to the former to keep consistency?